### PR TITLE
commit 1

### PR DESCRIPTION
**Pull request by dheerajsrinath**
Validating app download page. And respective app pages

### CODE REVIEW

File: src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.Iterator;
+import java.util.Set;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static com.codeborne.selenide.Selenide.open;
+import static com.codeborne.selenide.WebDriverRunner.getWebDriver;
+import static com.codeborne.selenide.WebDriverRunner.url;
+
+public class DownloadAppTest extends WebBaseTest {
```
Comment (akhil7000) [inline src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java]: Move this test to `FlipkartTest`

File: src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    FlipkartHomePage flipkartHomePage;
+    AppPage appPage;
```
Comment (akhil7000) [inline src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java]: make this local

File: src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    public void setup() {
+        open("https://www.flipkart.com/");
+        flipkartHomePage = new FlipkartHomePage().closePopup();
+    }
+
+    @ParameterizedTest
+    @CsvFileSource(resources = "/testDownloadAppTest.csv")
+    public void testAppDownloadValidation(String os, String appStoreUrl) {
+        appPage = flipkartHomePage.clickDownloadApp();
+        String appPageUrl = appPage.click(os).getUrl();
```
Comment (akhil7000) [inline src/test/java/com/training/web/tests/flipkart/DownloadAppTest.java]: use this directly in assert

File: src/main/java/com/training/web/pages/flipkart/AppPage.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+import static com.codeborne.selenide.Selenide.$x;
+import static com.codeborne.selenide.WebDriverRunner.getWebDriver;
+
+public class AppPage {
+    private String parentWindow=getWebDriver().getWindowHandle();
+    private String androidApp = "//img[contains(@src,'Play-Store')]";
+    private String iosApp= "//img[contains(@src,'App-Store')]";
+
+    public AppPage click(String os) {
```
Comment (akhil7000) [inline src/main/java/com/training/web/pages/flipkart/AppPage.java]: make this method more meaningful - `clickOs`

File: src/main/java/com/training/web/pages/flipkart/AppPage.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
+package com.training.web.pages.flipkart;
+
+import java.util.Iterator;
+import java.util.Set;
+
+import static com.codeborne.selenide.Selenide.$x;
+import static com.codeborne.selenide.WebDriverRunner.getWebDriver;
+
+public class AppPage {
+    private String parentWindow=getWebDriver().getWindowHandle();
```
Comment (akhil7000) [inline src/main/java/com/training/web/pages/flipkart/AppPage.java]: why is this used globally?

File: src/main/java/com/training/web/pages/flipkart/FlipkartHomePage.java [hunk 6]
```diff
@@ -13,6 +13,9 @@
     private String facebookIcon = "//u[text()='Facebook']";
     private String twitterIcon = "//a[@aria-label='Twitter']";
     private String youtubeIcon = "//paper-button[@aria-label='Subscribe']";
+    private String moreIcon = "//*[@class='_2gTTdy']";
+    private String downloadApp = "//div[text()='Download App']";
+    private String hassleFreeIcon = "//span[text()='Hassle - free Returns']";
```
Comment (akhil7000) [inline src/main/java/com/training/web/pages/flipkart/FlipkartHomePage.java]: This is violating the POM design - since this element is not in this page

**Review by akhil7000 — CHANGES_REQUESTED** [review]

File: src/main/java/com/training/web/pages/flipkart/AppPage.java (hunk 5 above)
Comment (dheerajsrinath) [reply src/main/java/com/training/web/pages/flipkart/AppPage.java]: this is used in 2 methods

Comment (akhil7000) [reply src/main/java/com/training/web/pages/flipkart/AppPage.java]: Which 2 methods? - I only see this in this method - `getUrl()`

File: src/main/java/com/training/web/pages/flipkart/ResultPage.java [hunk 7]
```diff
@@ -67,9 +67,8 @@ public ProductPage clickProduct(List<SelenideElement> productList, int itemNumbe
 
         $x(nextPageButton).shouldBe(visible);
         productList.get(itemNumber - 1).click();
-        String childWindow = null;
         String parentWindow = getWebDriver().getWindowHandle();
-
+        /*String childWindow = null;
```
Comment (akhil7000) [inline src/main/java/com/training/web/pages/flipkart/ResultPage.java]: remove the commented code

File: src/main/java/com/training/web/pages/flipkart/ResultPage.java [hunk 8]
```diff
@@ -78,7 +77,8 @@ public ProductPage clickProduct(List<SelenideElement> productList, int itemNumbe
                 childWindow = windows;
             }
         }
-        getWebDriver().switchTo().window(childWindow);
+        getWebDriver().switchTo().window(childWindow);*/
```
Comment (akhil7000) [inline src/main/java/com/training/web/pages/flipkart/ResultPage.java]: remove the commented code

**Review by akhil7000 — CHANGES_REQUESTED** [review]

**Review by akhil7000 — APPROVED** [review]

**Pull request merged**